File: frontend/lib/markdown.ts
const LINK = /\[([^\]]+)\]\(([^)]+)\)/g;
const CODE = /`([^`]+)`/g;
const BOLD = /\*\*([^*]+)\*\*/g;
const ITALIC = /(?<!\*)\*([^*]+)\*(?!\*)/g;

export function renderSafeMarkdown(text: string) {
  const source = (text || "").replace(/\r\n/g, "\n").replace(/\r/g, "\n").trim();
  if (!source) return "";
  const blocks: string[] = [];
  let listItems: string[] = [];
  let listKind = "";
  for (const raw of source.split("\n")) {
    const stripped = raw.trim();
    if (stripped.startsWith("- ") || stripped.startsWith("* ")) {
      if (listKind === "ol") {
        blocks.push(listHtml(listKind, listItems));
        listItems = [];
      }
      listKind = "ul";
      listItems.push(inline(stripped.slice(2)));
      continue;
    }
    if (/^\d+\.\s+/.test(stripped)) {
      if (listKind === "ul") {
        blocks.push(listHtml(listKind, listItems));
        listItems = [];
      }
      listKind = "ol";
      listItems.push(inline(stripped.replace(/^\d+\.\s+/, "")));
      continue;
    }
    if (listItems.length) {
      blocks.push(listHtml(listKind, listItems));
      listItems = [];
      listKind = "";
    }
    if (!stripped) continue;
    const heading = /^(#{1,3})\s+(.+)$/.exec(stripped);
    if (heading) {
      const level = Math.min(heading[1].length, 3) + 1;
      blocks.push(`<h${level}>${inline(heading[2])}</h${level}>`);
      continue;
    }
    blocks.push(`<p>${inline(stripped)}</p>`);
  }
  if (listItems.length) blocks.push(listHtml(listKind, listItems));
  return blocks.join("");
}

function listHtml(kind: string, items: string[]) {
  const tag = kind === "ol" ? "ol" : "ul";
  return `<${tag}>${items.map((item) => `<li>${item}</li>`).join("")}</${tag}>`;
}

function inline(text: string) {
  let escaped = escapeHtml(text);
  escaped = escaped.replace(LINK, (_all, label: string, href: string) => {
    const safe = safeHref(href);
    return safe ? `<a href="${escapeAttr(safe)}" target="_blank" rel="noopener noreferrer">${label}</a>` : label;
  });
  escaped = escaped.replace(CODE, (_all, value: string) => `<code>${value}</code>`);
  escaped = escaped.replace(BOLD, (_all, value: string) => `<strong>${value}</strong>`);
  escaped = escaped.replace(ITALIC, (_all, value: string) => `<em>${value}</em>`);
  return escaped;
}

function escapeHtml(value: string) {
  return value.replace(/[&<>"']/g, (ch) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[ch] || ch));
}

function escapeAttr(value: string) {
  return escapeHtml(value).replace(/"/g, "&quot;");
}

function safeHref(url: string) {
  const text = (url || "").trim();
  const lower = text.toLowerCase();
  if (lower.startsWith("javascript:") || lower.startsWith("data:") || lower.startsWith("vbscript:") || lower.startsWith("file:")) return "";
  if (lower.startsWith("mailto:")) {
    const address = text.slice(7).trim();
    return /^[^\s@<>]+@[^\s@<>]+\.[^\s@<>]+$/.test(address) ? `mailto:${address}` : "";
  }
  try {
    const parsed = new URL(text.includes("://") ? text : `https://${text}`);
    const host = (parsed.hostname || "").toLowerCase();
    if ((parsed.protocol !== "http:" && parsed.protocol !== "https:") || parsed.username) return "";
    if (!host || host === "javascript" || host === "vbscript" || host === "data") return "";
    return text.includes("://") ? text : `https://${text}`;
  } catch {
    return "";
  }
}

File: backend/app/core/markdown.py
import re
from html import escape

from app.core.network_safety import safe_public_url

LINK = re.compile(r"\[([^\]]+)\]\(([^)]+)\)")
CODE = re.compile(r"`([^`]+)`")
BOLD = re.compile(r"\*\*([^*]+)\*\*")
ITALIC = re.compile(r"(?<!\*)\*([^*]+)\*(?!\*)")
HEADING = re.compile(r"^(#{1,3})\s+(.+)$")


def render_safe_markdown(text: str) -> str:
    source = (text or "").replace("\r\n", "\n").replace("\r", "\n").strip()
    if not source:
        return ""
    blocks: list[str] = []
    list_items: list[str] = []
    list_kind = ""
    for raw in source.split("\n"):
        line = raw.rstrip()
        stripped = line.strip()
        if stripped.startswith("- ") or stripped.startswith("* "):
            if list_kind == "ol":
                blocks.append(_list_html(list_kind, list_items))
                list_items = []
            list_kind = "ul"
            list_items.append(_inline(stripped[2:]))
            continue
        if re.match(r"^\d+\.\s+", stripped):
            if list_kind == "ul":
                blocks.append(_list_html(list_kind, list_items))
                list_items = []
            list_kind = "ol"
            list_items.append(_inline(re.sub(r"^\d+\.\s+", "", stripped)))
            continue
        if list_items:
            blocks.append(_list_html(list_kind, list_items))
            list_items = []
            list_kind = ""
        if not stripped:
            continue
        heading = HEADING.match(stripped)
        if heading:
            level = min(len(heading.group(1)), 3) + 1
            blocks.append(f"<h{level}>{_inline(heading.group(2))}</h{level}>")
            continue
        blocks.append(f"<p>{_inline(stripped)}</p>")
    if list_items:
        blocks.append(_list_html(list_kind, list_items))
    return "".join(blocks)


def _list_html(kind: str, items: list[str]) -> str:
    tag = "ol" if kind == "ol" else "ul"
    return f"<{tag}>{''.join(f'<li>{item}</li>' for item in items)}</{tag}>"


def _inline(text: str) -> str:
    escaped = escape(text)

    def link(match: re.Match[str]) -> str:
        label = match.group(1)
        href = _safe_href(match.group(2))
        if not href:
            return label
        return f'<a href="{escape(href, quote=True)}" target="_blank" rel="noopener noreferrer">{label}</a>'

    escaped = LINK.sub(link, escaped)
    escaped = CODE.sub(lambda match: f"<code>{match.group(1)}</code>", escaped)
    escaped = BOLD.sub(lambda match: f"<strong>{match.group(1)}</strong>", escaped)
    escaped = ITALIC.sub(lambda match: f"<em>{match.group(1)}</em>", escaped)
    return escaped


def _safe_href(url: str) -> str | None:
    text = (url or "").strip()
    lower = text.lower()
    if lower.startswith(("javascript:", "data:", "vbscript:", "file:")):
        return None
    if lower.startswith("mailto:"):
        address = text.split(":", 1)[1].strip()
        return f"mailto:{address}" if re.fullmatch(r"[^\s@<>]+@[^\s@<>]+\.[^\s@<>]+", address) else None
    candidate = text if "://" in text else f"https://{text}"
    return safe_public_url(candidate)
